src/Microsoft.Identity.Web/Resource/AadIssuerValidator.cs

```diff
@@ -10,198 +10,198 @@
+				throw new ArgumentNullException(nameof(actualIssuer));
+			}
+
+			if (securityToken == null)
+			{
+				throw new ArgumentNullException(nameof(securityToken));
+			}
+
+			if (validationParameters == null)
+			{
+				throw new ArgumentNullException(nameof(validationParameters));
+			}
+
+			string tenantId = GetTenantIdFromToken(securityToken);
+			if (string.IsNullOrWhiteSpace(tenantId))
+			{
+				throw new SecurityTokenInvalidIssuerException(IDWebErrorMessage.TenantIdClaimNotPresentInToken);
+			}
+
+			if (validationParameters.ValidIssuers != null)
+			{
+				foreach (var validIssuerTemplate in validationParameters.ValidIssuers)
+				{
+					if (IsValidIssuer(validIssuerTemplate, tenantId, actualIssuer))
+					{
+						return actualIssuer;
+					}
+				}
+			}
+
+			if (IsValidIssuer(validationParameters.ValidIssuer, tenantId, actualIssuer))
+			{
+				return actualIssuer;
+			}
+
+			// If a valid issuer is not found, throw
+			throw new SecurityTokenInvalidIssuerException(
+				string.Format(
+					CultureInfo.InvariantCulture,
+					IDWebErrorMessage.IssuerDoesNotMatchValidIssuers,
+					actualIssuer));
+		}
+
+		private bool IsValidIssuer(string validIssuerTemplate, string tenantId, string actualIssuer)
+		{
+			if (string.IsNullOrEmpty(validIssuerTemplate))
+			{
+				return false;
+			}
+
+			try
+			{
+				Uri issuerFromTemplateUri = new Uri(validIssuerTemplate.Replace("{tenantid}", tenantId));
+				Uri actualIssuerUri = new Uri(actualIssuer);
+
+				// Template authority is in the aliases
+				return _issuerAliases.Contains(issuerFromTemplateUri.Authority) &&
+					   // "iss" authority is in the aliases
+					   _issuerAliases.Contains(actualIssuerUri.Authority) &&
+					  // Template authority ends in the tenant ID
+					  IsValidTidInLocalPath(tenantId, issuerFromTemplateUri) &&
+					  // "iss" ends in the tenant ID
+					  IsValidTidInLocalPath(tenantId, actualIssuerUri);
+			}
+			catch
+			{
+				// if something faults, ignore
+			}
+
+			return false;
+		}
+
+		private static bool IsValidTidInLocalPath(string tenantId, Uri uri)
+		{
+			string trimmedLocalPath = uri.LocalPath.Trim('/');
+			return trimmedLocalPath == tenantId || trimmedLocalPath == $"{tenantId}/v2.0";
+		}
+
+		/// <summary>Gets the tenant ID from a token.</summary>
+		/// <param name="securityToken">A JWT token.</param>
+		/// <returns>A string containing the tenant ID, if found or <see cref="string.Empty"/>.</returns>
+		/// <remarks>Only <see cref="JwtSecurityToken"/> and <see cref="JsonWebToken"/> are acceptable types.</remarks>
+		private static string GetTenantIdFromToken(SecurityToken securityToken)
+		{
+			if (securityToken is JwtSecurityToken jwtSecurityToken)
+			{
+				if (jwtSecurityToken.Payload.TryGetValue(ClaimConstants.Tid, out object? tid))
+				{
+					return (string)tid;
+				}
+
+				jwtSecurityToken.Payload.TryGetValue(ClaimConstants.TenantId, out object? tenantId);
+				if (tenantId != null)
+				{
+					return (string)tenantId;
```

Comment:
in case someone is trying to break things, the cast would be safer as `tenantId as string;`